### PR TITLE
Tweak flaky meta data save test

### PR DESCRIPTION
**Pull request by lsinger**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

The `test_order_save_is_not_called_frequently_on_meta_save` test was flaky a few times before, most likely caused by a timing issue (cf. [`should_save_after_meta_change`](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php#L2956)). This PR adds the possibility of `time()` having advanced during the five calls to `save_meta_data`. 

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. ensure CI passes
2. ensure the updated test looks good still

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

**Comment by Konamiman** [comment]
@lsinger While the fix works, it seems somewhat hacky. What about modifying `OrdersTableDataStore::should_save_after_meta_change` to use the `LegacyProxy` (which the class is receiving in its `init` method already) to retrieve the current time instead of directly using `current_time`? Then in the test we can set the current time to whatever value we want and test the two cases: before and after the order modification time. Also this has the added benefit that the test reads better.

**Comment by lsinger** [comment]
> @lsinger While the fix works, it seems somewhat hacky. What about modifying `OrdersTableDataStore::should_save_after_meta_change` to use the `LegacyProxy` (which the class is receiving in its `init` method already) to retrieve the current time instead of directly using `current_time`? Then in the test we can set the current time to whatever value we want and test the two cases: before and after the order modification time. Also this has the added benefit that the test reads better.

@Konamiman I agree that the fix doesn't feel perfect. I'd prefer to be a bit hacky-ish in a test, though, over doing the same in production code just so that a test runs better. 

And even though `LegacyProxy` is already available, I'd prefer we don't increase coupling to this legacy (?) class even further. Isn't the idea of the `LegacyProxy` to provide a somewhat temporary bridge to Woo's legacy functions that would ideally be phased out over time? Maybe I'm misunderstanding the intent behind this class ...? 🤔

**Comment by Konamiman** [comment]
The purpose of the `LegacyProxy` is _precisely_ to ease testing. 😄

For any new code we use the dependency injection container, which allows us to easily replace dependencies in tests. But for the cases where we can't do that (old/non-WC functions, old not DI-ed classes in `includes`) the legacy proxy can be (and is) used in production to make these "legacy" code paths easily replaceable as if they were new code. Also this is done already in other parts of the same class (search for instances of `$this->legacy_proxy->call_function`).

>  Isn't the idea of the LegacyProxy to provide a somewhat temporary bridge to Woo's legacy functions that would ideally be phased out over time?

Yes, but provided that we have a proper replacement in the form of a DI-ed new class. In this case we are dealing with the WordPress core `current_time` function, so there's little we can do (other than creating a new utility class with a method that provides the current time - maybe overkill, or maybe not).

**Comment by lsinger** [comment]
@Konamiman thank you for that clarification! I just pushed an update that I believe does what you're alluding to.

File: plugins/woocommerce/tests/php/src/Internal/DataStores/Orders/OrdersTableDataStoreTests.php [hunk 1]
```diff
@@ -3085,6 +3085,21 @@ function () use ( &$count ) {
...
+		 * less than a second of time passes while the following saves happen.
+		 */
+		$datetime = new DateTime( 'now', new DateTimeZone( 'UTC' ) );
+		$now      = $datetime->format( 'Y-m-d H:i:s' );
+		$this->reset_legacy_proxy_mocks();
+		$this->register_legacy_proxy_function_mocks(
+			array(
+				'current_time' => function( $type, $gmt ) {
+					return $now;
+				},
```
Comment (Konamiman) [inline plugins/woocommerce/tests/php/src/Internal/DataStores/Orders/OrdersTableDataStoreTests.php]: This is incorrect: `$now` is interpreted as a variable defined inside the function and thus the value actually returned is undefined (probably `null`). It should have an `use` clause:

```php
'current_time' => function( $type, $gmt ) use ( $now ) {
	return $now;
},
```

Or alternatively, the new lambda syntax introduced in PHP 7.4 could be used which is more concise (and doesn't require `use`s):

```php
'current_time' => fn( $type, $gmt ) => $now
```

**Review by Konamiman — CHANGES_REQUESTED** [review]

File: plugins/woocommerce/tests/php/src/Internal/DataStores/Orders/OrdersTableDataStoreTests.php (hunk 1 above)
Comment (lsinger) [reply plugins/woocommerce/tests/php/src/Internal/DataStores/Orders/OrdersTableDataStoreTests.php]: Great catch, thank you @Konamiman! Looking into it further I've discovered that the `current_time` mock never gets called in the first place. I'll explore this some more.

Comment (lsinger) [reply plugins/woocommerce/tests/php/src/Internal/DataStores/Orders/OrdersTableDataStoreTests.php]: Addressed this. ✅

**Review by Konamiman — APPROVED** [review]

**Pull request merged**